### PR TITLE
patch2pr: init at 0.22.0

### PR DESCRIPTION
**Pull request by katrinafyi**
## Description of changes

https://github.com/bluekeyes/patch2pr

_patch2pr_ creates pull requests from patch files. it uses the Github API to create PRs without needing a repo checkout. 

the package is a fairly simple go package.

testing functionality will need a GITHUB_TOKEN. assuming you have a nixpkgs fork and a token with write permissions on this fork, you can test this with:
```
wget -O test.patch https://github.com/NixOS/nixpkgs/pull/291104.patch
GITHUB_TOKEN=github_pat_...   patch2pr -repository NixOS/nixpkgs -fork -patch-base 2ea79083c54ea7a665562c1c41b0e18d9a9ff72e -json -no-pull-request ./test.patch
```
this should make a branch called "patch2pr" in your fork with this pull request applied. with more permissions, you can test without -no-pull-request.

there is also a simple testVersion:
```
nix build .#patch2pr.tests.patch2pr-version
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/pa/patch2pr/package.nix [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{ lib
+, buildGoModule
+, fetchFromGitHub
+, testers
+, patch2pr
+}:
+
+buildGoModule (self: {
```
Comment (AndersonTorres) [inline pkgs/by-name/pa/patch2pr/package.nix:8]:
```suggestion
buildGoModule rec {
```

buildGoModule does not recognize finalAttrs design pattern yet.

File: pkgs/by-name/pa/patch2pr/package.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    owner = "bluekeyes";
+    repo = "patch2pr";
+    rev = "v${self.version}";
+    hash = "sha256-mWpCyZ6T9q6wL5PyRRptWL9sf+QPydU1onqmjLQxiZk=";
+  };
+
+  vendorHash = "sha256-vUDs5GKjYwuzSpz5JKakzqs9sZ/XrscxfrS+bFQ7BZ8=";
+
+  passthru.tests.patch2pr-help = testers.testVersion
+    {
```
Comment (AndersonTorres) [inline pkgs/by-name/pa/patch2pr/package.nix:22]:
```suggestion
  passthru.tests.patch2pr-help = testers.testVersion {
```

File: pkgs/by-name/pa/patch2pr/package.nix [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      package = patch2pr;
+      command = "patch2pr --help";
+      version = "patch2pr";
+    };
+
+  meta = {
+    description = "Create pull requests from patches without cloning the repository";
+    homepage = "https://github.com/bluekeyes/patch2pr";
+    changelog = "https://github.com/bluekeyes/patch2pr/releases/tag/v${self.version}";
```
Comment (AndersonTorres) [inline pkgs/by-name/pa/patch2pr/package.nix:31]:
```suggestion
    changelog = "https://github.com/bluekeyes/patch2pr/releases/tag/v${src.rev}";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by katrinafyi** [comment]
@AndersonTorres thanks! changes made to use rec instead and adjust style.

File: pkgs/by-name/pa/patch2pr/package.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+}:
+
+buildGoModule rec {
+  pname = "patch2pr";
+  version = "0.21.0";
+
+  src = fetchFromGitHub {
+    owner = "bluekeyes";
+    repo = "patch2pr";
+    rev = "v0.21.0";
```
Comment (AndersonTorres) [inline pkgs/by-name/pa/patch2pr/package.nix:15]:
```suggestion
    rev = "v${version}";
```

File: pkgs/by-name/pa/patch2pr/package.nix [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    version = "patch2pr";
+  };
+
+  meta = with lib; {
+    description = "Create pull requests from patches without cloning the repository";
+    homepage = "https://github.com/bluekeyes/patch2pr";
+    changelog = "https://github.com/bluekeyes/patch2pr/releases/tag/v${src.rev}";
+    license = licenses.mit;
+    maintainers = with maintainers; [ katrinafyi ];
+    platforms = platforms.all;
```
Comment (AndersonTorres) [inline pkgs/by-name/pa/patch2pr/package.nix:33]:
```suggestion
```

Go is not built for all platforms.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by katrinafyi** [comment]
that was a mistake with nurl, very sorry about that!

**Comment by katrinafyi** [comment]
I have updated the package to 0.22.0. This is a separate commit so nixpkgs records multiple versions, just in case it's useful.

**Review by AndersonTorres — APPROVED** [review]

**Pull request merged**